lua/pixelui/core/cl_color.lua

```diff
@@ -45,96 +28,85 @@ function PIXEL.Hue2RGB(p, q, t)
-    local b = col.b / 255
-    local max, min = math.max(r, g, b), math.min(r, g, b)
-    b = max + min
-    local h = b / 2
-    if max == min then return 0, 0, h end
-    local s, l = h, h
-    local d = max - min
-    s = l > .5 and d / (2 - b) or d / b
-    if max == r then h = (g - b) / d + (g < b and 6 or 0)
-    elseif max == g then h = (b - r) / d + 2
-    elseif max == b then h = (r - g) / d + 4
+do
+    local format = string.format
+    function PIXEL.DecToHex(dec, zeros)
+        return format("%0" .. (zeros or 2) .. "x", dec)
     end
-    return h * .16667, s, l
-end
 
-function PIXEL.DecToHex(d, zeros)
-    return string.format("%0" .. (zeros or 2) .. "x", d)
+    local max = math.max
+    local min = math.min
+    function PIXEL.ColorToHex(color)
+        return format("#%02X%02X%02X",
+            max(min(color.r, 255), 0),
+            max(min(color.g, 255), 0),
+            max(min(color.b, 255), 0)
+        )
+    end
 end
 
-function PIXEL.RGBToHex(color)
-    return "#" ..
-        PIXEL.DecToHex(math.max(math.min(color.r, 255), 0)) ..
-        PIXEL.DecToHex(math.max(math.min(color.g, 255), 0)) ..
-        PIXEL.DecToHex(math.max(math.min(color.b, 255), 0))
+do
+    local match = string.match
+    local tonumber = tonumber
+
+    function PIXEL.HexToColor(hex)
+        local r, g, b = match(hex, "#(..)(..)(..)")
+        return createColor(
+            tonumber(r, 16),
+            tonumber(g, 16),
+            tonumber(b, 16)
+        )
+    end
 end
 
-function PIXEL.HexToRGB(hex)
-    hex = hex:gsub("#", "")
+do
+    local curTime = CurTime
+    local hsvToColor = HSVToColor
 
-    if (#hex == 3) then
-        local r = hex:sub(1, 1)
-        local g = hex:sub(2, 2)
-        local b = hex:sub(3, 3)
+    local lastUpdate = 0
+    local lastCol = createColor(0, 0, 0)
 
-        return Color(
-        tonumber("0x" .. r .. r),
-        tonumber("0x" .. g .. g),
-        tonumber("0x" .. b .. b)
-        )
-    end
+    function PIXEL.GetRainbowColor()
+        local time = curTime()
+        if lastUpdate == time then return lastCol end
 
-    return Color(
-        tonumber("0x" .. hex:sub(1, 2)),
-        tonumber("0x" .. hex:sub(3, 4)),
-        tonumber("0x" .. hex:sub(5, 6))
-    )
+        lastUpdate = time
+        lastCol = hsvToColor((time * 50) % 360, 1, 1)
+
+        return lastCol
+    end
 end
 
-local lastUpdate = 0
-local lastCol = Color(0, 0, 0)
-local HSVToColor = HSVToColor
-function PIXEL.GetRainbowColor()
-    local time = CurTime()
-    if lastUpdate == time then return lastCol end
+do
+    local colorToHSL = ColorToHSL
+
+    function PIXEL.IsColorLight(col)
+        local _, _, lightness = colorToHSL(col)
+        return lightness >= .5
+    end
+end
 
-    lastUpdate = time
-    lastCol = HSVToColor((time * 50) % 360, 1, 1)
+function PIXEL.LerpColor(t, from, to)
+    return createColor(from.r, from.g, from.b, from.a):Lerp(t, to)
+end
 
-    return lastCol
+function PIXEL.IsColorEqualTo(from, to)
```

Comment:
This looks dreadful, would it not be better to store this as a number, or some sort of hash then compare the hash? Four different comparissions will be slow if called frequently.

Reply:
I can't imagine any use cases of this in a draw hook, your suggestion seems a little over complicated for something that's likely rarely going to be called.